src/Interpreters/Context.cpp

```diff
@@ -912,6 +912,23 @@ String Context::getFilesystemCacheUser() const
     return shared->filesystem_cache_user;
 }
 
+DatabaseAndTable Context::getOrCacheStorage(const StorageID & id, std::function<DatabaseAndTable()> f, std::optional<Exception> * exception) const
+{
+    if (auto storage_id = storage_cache.find(id); storage_id != storage_cache.end())
+    {
+        DatabaseAndTable storage = DatabaseCatalog::instance().tryGetByUUID(storage_id->uuid);
+        if (exception && !storage.second)
+            exception->emplace(Exception(ErrorCodes::UNKNOWN_TABLE, "Table {} does not exist anymore - maybe it was dropped", id.getNameForLogs()));
```

Comment:
Currently, a query will fail if it requests a table the second time, but the table was dropped after the first get. But we can improve it later

Reply:
I don't think we need to improve it - if table was dropped concurrently then query is better to be interrupted than trying to reposes the table from drop queue - at this point we don't even know who is holding it from physical deletion...

Reply:
We do know, see `addStorageHolder`